pkg/adsc/delta.go

```diff
@@ -0,0 +1,703 @@
+		errChan:         make(chan error, 10),
+		deltaXDSUpdates: make(chan *discovery.DeltaDiscoveryResponse, 100),
+		received:        map[string]*discovery.DeltaDiscoveryResponse{},
+		mutex:           sync.RWMutex{},
+		resourceCache:   sync.Map{},
+	}
+	for _, o := range opts {
+		o(c)
+	}
+	return c
+}
+
+func NewDeltaWithBackoffPolicy(discoveryAddr string, config *DeltaADSConfig, backoffPolicy backoff.BackOff, opts ...Option) *Client {
+	if config == nil {
+		config = &DeltaADSConfig{}
+	}
+	delta := NewDelta(discoveryAddr, config, opts...)
+	delta.cfg.BackoffPolicy = backoffPolicy
+	return delta
+}
+
+func typeName[T proto.Message]() string {
+	ft := new(T)
+	return resource.APITypePrefix + string((*ft).ProtoReflect().Descriptor().FullName())
+}
+
+// Register registers a handler for a type which is reflected by the proto message.
+func Register[T proto.Message](f func(ctx HandlerContext, res T, event Event)) Option {
+	return func(c *Client) {
+		c.handlers[typeName[T]()] = func(ctx HandlerContext, res proto.Message, event Event) {
+			f(ctx, res.(T), event)
+		}
+	}
+}
+
+// Watch registers an initial watch for a type based on the type reflected by the proto message.
+func Watch[T proto.Message](resourceName string) Option {
+	return initWatch(typeName[T](), resourceName)
+}
+
+func initWatch(typeURL string, resourceName string) Option {
+	return func(c *Client) {
+		if resourceName == "*" {
+			// Normalize to allow both forms
+			resourceName = ""
+		}
+		key := resourceKey{
+			Name:    resourceName,
+			TypeURL: typeURL,
+		}
+		existing, f := c.tree[key]
+		if f {
+			// We are watching directly now, so erase any parents
+			existing.Parents = nil
+		} else {
+			c.tree[key] = resourceNode{
+				Parents:  make(keySet),
+				Children: make(keySet),
+			}
+		}
+		c.initialWatches = append(c.initialWatches, key)
+	}
+}
+
+func (c *Client) handleRecv() {
+	// We connected, so reset the backoff
+	if c.cfg.BackoffPolicy != nil {
+		c.cfg.BackoffPolicy.Reset()
+	}
+	for {
+		deltaLog.Infof("Start Recv for node %v", c.nodeID)
+		msg, err := c.xdsClient.Recv()
+		if err != nil {
+			deltaLog.Infof("Connection closed for node %v with err: %v", c.nodeID, err)
+			select {
+			case c.errChan <- err:
+			default:
+			}
+			// if 'reconnect' enabled - schedule a new Run
+			if c.cfg.BackoffPolicy != nil {
+				time.AfterFunc(c.cfg.BackoffPolicy.NextBackOff(), c.reconnect)
+			} else {
+				c.Close()
+				c.WaitClear()
+				c.deltaXDSUpdates <- nil
+				close(c.errChan)
+			}
+			return
+		}
+		deltaLog.Infof("Received response: %s", msg.TypeUrl)
+		if err := c.handleDeltaResponse(msg); err != nil {
+			deltaLog.Infof("Handle response %s failed: %v", msg.TypeUrl, err)
+			c.Close()
+			return
+		}
+	}
+}
+
+func (c *Client) handleDeltaResponse(d *discovery.DeltaDiscoveryResponse) error {
+	var rejects []error
+	allAdds := map[string]set.Set[string]{}
+	allRemoves := map[string]set.Set[string]{}
+	ctx := &handlerContext{}
+	defer func() {
```

Comment:
For debugType, we need to record and use the wait function to wait for resources. Otherwise, we won't know when the handling of the subscribed resources has ended.